### PR TITLE
Refactor merge & DML test helpers to allow reusability

### PR DESCRIPTION
**Pull request by johanl-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
This change is a plain refactor of our test helpers, gathering merge and DML test helpers outside of the existing test suites into traits that can be reused in separate suites.

## How was this patch tested?
N/A

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+    s"""
+       |MERGE INTO $target
+       |USING $source
+       |ON $condition
+       |WHEN MATCHED THEN UPDATE SET $update
+       |WHEN NOT MATCHED THEN INSERT $insert
+      """.stripMargin
+  }
+
+  override def executeMerge(
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala:90]: Could we add the `protected` access modified for `executeMerge`?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala [hunk 2]
```diff
@@ -358,3 +363,48 @@ trait DeltaTestUtilsForTempViews
...
+  extends DeltaTestUtilsBase  with BeforeAndAfterEach {
+  self: SharedSparkSession =>
+
+  protected var tempDir: File = _
+
+  protected var deltaLog: DeltaLog = _
+
+  protected def tempPath: String = tempDir.getCanonicalPath
+
+  override def beforeEach(): Unit = {
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala:381]: While you're changing this, please see if you can set the access modifier to `protected` for afterEach and beforeEach. I was always annoyed by it not having the correct access modifier

**Review by fred-db — COMMENTED** [review]
Some small nits, otherwise LGTM. Thanks for the cleanup! :)

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala (hunk 2 above)
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala]: Done :+1:

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoSQLSuite.scala [hunk 3]
```diff
@@ -30,46 +30,13 @@ import org.apache.spark.sql.functions.udf
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types.{IntegerType, StructField, StructType}
 
-class MergeIntoSQLSuite extends MergeIntoSuiteBase  with MergeIntoNotMatchedBySourceSuite
+class MergeIntoSQLSuite extends MergeIntoSuiteBase
+  with MergeIntoSQLTestUtils  with MergeIntoNotMatchedBySourceSuite
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoSQLSuite.scala:34]: nit: 1 space indent here
```suggestion
  with MergeIntoSQLTestUtils with MergeIntoNotMatchedBySourceSuite
```

File: spark/src/test/scala/org/apache/spark/sql/delta/DeleteSuiteBase.scala [hunk 4]
```diff
@@ -17,61 +17,22 @@
...
 import org.apache.spark.sql.functions.struct
 import org.apache.spark.sql.test.SharedSparkSession
-import org.apache.spark.sql.types.{IntegerType, StructField, StructType}
-import org.apache.spark.util.Utils
+import org.apache.spark.sql.types.StructType
 
 abstract class DeleteSuiteBase extends QueryTest
   with SharedSparkSession
-  with BeforeAndAfterEach  with DeltaTestUtilsForTempViews {
+  with DeltaDMLTestUtils  with DeltaTestUtilsForTempViews {
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeleteSuiteBase.scala:30]: nit: one space indent here
```suggestion
  with DeltaDMLTestUtils with DeltaTestUtilsForTempViews {
```

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala [hunk 5]
```diff
@@ -358,3 +363,48 @@ trait DeltaTestUtilsForTempViews
     }
   }
 }
+
+/**
+ * Trait collecting helper methods for DML tests e.p. creating a test table for each test and
+ * cleaning it up after each test.
+ */
+trait DeltaDMLTestUtils
+  extends DeltaTestUtilsBase  with BeforeAndAfterEach {
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaTestUtils.scala:372]:
```suggestion
  extends DeltaTestUtilsBase with BeforeAndAfterEach {
```

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala [hunk 6]
```diff
@@ -21,15 +21,15 @@ import java.util.Locale
...
 
 import org.apache.spark.sql._
 import org.apache.spark.sql.catalyst.plans.Inner
 import org.apache.spark.sql.catalyst.plans.logical.{Assignment, DeltaMergeIntoClause, Join}
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.types.StructType
 
-class MergeIntoScalaSuite extends MergeIntoSuiteBase  with MergeIntoNotMatchedBySourceSuite
+class MergeIntoScalaSuite extends MergeIntoSuiteBase
+  with MergeIntoScalaTestUtils  with MergeIntoNotMatchedBySourceSuite
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoScalaSuite.scala:32]:
```suggestion
  with MergeIntoScalaTestUtils with MergeIntoNotMatchedBySourceSuite
```

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala (hunk 1 above)
Comment (fred-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala:90]: Also where we override it, e.g. here

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+      target: String,
+      source: String,
+      condition: String,
+      update: String,
+      insert: String): Unit =
+    sql(basicMergeStmt(target, source, condition, update, insert))
+
+  override def executeMerge(
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala:98]: or here
```suggestion
  override protected def executeMerge(
```

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+      src: String,
+      cond: String,
+      clauses: MergeClause*): Unit =
+    sql(s"MERGE INTO $tgt USING $src ON $cond\n" + clauses.map(_.sql).mkString("\n"))
+}
+
+trait MergeIntoScalaTestUtils extends MergeIntoTestUtils {
+  self: SharedSparkSession =>
+
+  override def executeMerge(
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala:109]: or here
```suggestion
  override protected def executeMerge(
```

**Review by vkorukanti — APPROVED** [review]

File: spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala (hunk 1 above)
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/MergeIntoTestUtils.scala]: Good catch, I though I had it covered the first time

**Review by fred-db — APPROVED** [review]
LGTM!

**Pull request closed without merging**